src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,296 @@
+import * as _ from 'lodash';
+import * as grpc from 'grpc';
+import * as Bluebird from 'bluebird';
+import { EventEmitter } from 'events';
+
+import {
+	createInsecureCredentials,
+	createOrgIdMetadata,
+	Direction,
+	EntryAdapter,
+	IGrpcClientAsync,
+	promisifyClient,
+	PusherClient,
+	PushRequest,
+	QuerierClient,
+	QueryRequest,
+	QueryResponse,
+	status,
+	StreamAdapter,
+	TailRequest,
+	TailResponse,
+	Timestamp,
+} from 'loki-grpc-client';
+import { errors } from '@balena/pinejs';
+import { LOKI_HOST, LOKI_PORT } from '../../../lib/config';
+import {
+	DeviceLog,
+	DeviceLogsBackend,
+	LogContext,
+	LogWriteContext,
+	Subscription,
+} from '../struct';
+import { captureException } from '../../../platform/errors';
+
+const { BadRequestError } = errors;
+
+const MIN_BACKOFF = 100;
+const MAX_BACKOFF = 10 * 1000;
+const VERSION = 1;
+
+function createTimestampFromDate(date = new Date()) {
+	const timestamp = new Timestamp();
+	timestamp.fromDate(date);
+	return timestamp;
+}
+
+function createTimestampFromMilliseconds(milliseconds: number) {
+	const MS_TO_NANOS = 1000000;
+	const seconds = Math.floor(milliseconds / 1000);
+	const nanos = (milliseconds - seconds * 1000) * MS_TO_NANOS;
+
+	const timestamp = new Timestamp();
+	timestamp.setSeconds(seconds);
+	timestamp.setNanos(nanos);
+	return timestamp;
+}
+
+export class LokiBackend implements DeviceLogsBackend {
+	private subscriptions: EventEmitter;
+	private querier: QuerierClient;
+	private pusher: IGrpcClientAsync;
+	private ready: boolean;
+	private tailCalls: Map<string, grpc.ClientReadableStream<TailResponse>>;
```

Comment:
`grpc.ClientReadableStream` seems like it'd make sense to export alongside `QuerierClient` since the semantic type we care about is `ReturnType<QuerierClient['tail']>`

Reply:
Updated